server/src/main/java/org/candlepin/controller/Entitler.java

```diff
@@ -205,6 +250,102 @@ public void adjustEntitlementQuantity(Consumer consumer, Entitlement ent,
+        Set<Product> providedProducts = new HashSet<Product>();
+        Date now = new Date();
+        List<String> devProductIds = new ArrayList<String>();
+        Product skuProd = null;
+
+        devProductIds.add(sku);
+        for (ConsumerInstalledProduct ip : consumer.getInstalledProducts()) {
+            devProductIds.add(ip.getProductId());
+        }
+        List<Product> prods = productAdapter.getProductsByIds(consumer.getOwner(), devProductIds);
+        if (prods.size() != devProductIds.size()) {
+            // determine exception message
+            produceMissingDevProductException(devProductIds, prods);
+        }
+        for (Product prod : prods) {
+            prod.setOwner(consumer.getOwner());
+            for (ProductContent pc : prod.getProductContent()) {
+                pc.getContent().setOwner(consumer.getOwner());
+            }
+            if (sku.equals(prod.getId())) {
+                if (StringUtils.isEmpty(prod.getAttributeValue("support_level"))) {
+                    // if there is no SLA, apply the default
+                    prod.setAttribute("support_level", this.DEFAULT_DEV_SLA);
+                }
+                skuProd = prod;
+            }
+            else {
+                providedProducts.add(prod);
+            }
+            productCurator.createOrUpdate(prod);
+        }
+
+        Date then = new Date(now.getTime() + getPoolInterval(skuProd));
+        Pool p = new Pool(consumer.getOwner(), skuProd, providedProducts, 1L, now, then, "", "", "");
+        p.setAttribute(Pool.DEVELOPMENT_POOL_ATTRIBUTE, "true");
+        p.setAttribute(Pool.REQUIRES_CONSUMER_ATTRIBUTE, consumer.getUuid());
+        return p;
+    }
+
+    /**
+     *
+     * @param prodIds List of Ids we expect in the product list. The first is the sku.
+     * @param products
+     * @return
+     * @throws ForbiddenException
+     */
+    protected void produceMissingDevProductException(List<String> prodIds, List<Product> products)
+            throws ForbiddenException {
+        StringBuffer message = new StringBuffer();
+        Map<String, Product> prodMap = new HashMap<String, Product>();
+        for (Product prod : products) {
+            prodMap.put(prod.getId(), prod);
+        }
+        for (int i = 0; i < prodIds.size(); i++) {
+            if (prodMap.get(prodIds.get(i)) == null) {
+                if (i == 0) {
+                    message.append(i18n.tr("The sku product ''{0}'' for this development " +
+                            "unit does not exist in the repository. ",
+                            prodIds.get(i)));
+                }
+                else {
+                    message.append(i18n.tr("The installed product ''{0}'' for this development " +
+                            "unit does not exist in the repository. ",
+                            prodIds.get(i)));
+                }
+            }
+        }
+        throw new ForbiddenException(message.toString());
+    }
+
```

Comment:
This could produce a very large message, containing some duplication with "The installed product {0} for this development unit does not exist in the repository." for each product. Would using 'repository' here be a little confusing?

Also, IMO we should throw a separate exception for unknown SKU vs unknown installed products.

Why not just loop through the products and remove the product_id from the list of prodIds if found. Any left over are products that were not found. The check that list to see if the sku was found. if not, throw an exception. If the SKU is not found, autobind failed.

Then if the sku was found and there are products remaining in the list, throw an exception with a single message and perhaps comma separated list of installed product ids.